vmdk_plugin/utils/fs/device_watcher_windows.go

```diff
@@ -0,0 +1,108 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Device watching logic for Windows OS.
+
+package fs
+
+import (
+	"fmt"
+	"os/exec"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+const (
+	devChangeTimeout = 1 // Number of seconds to wait for a Win32_DeviceChangeEvent
+
+	// PowerShell script to register and wait for a Win32_DeviceChangeEvent.
+	// The script blocks until a Win32_DeviceChangeEvent occurs or the event times out.
+	devChangeWaitScript = `
+		Register-WmiEvent -Class Win32_DeviceChangeEvent -SourceIdentifier DeviceChangeEvent
+		$event = Wait-Event -SourceIdentifier DeviceChangeEvent -Timeout %d
+		If ($event) {
+			Write-Host "DeviceChangeEvent"
+		}
+		Else {
+			Write-Host "EventTimedOut"
+		}
+	`
+)
+
+// DeviceWatcher represents a device change watcher.
+type DeviceWatcher struct {
+	Event      chan string   // chan for emitting events
+	Error      chan error    // chan for emitting errors
+	stop       bool          // flag to initiate watcher termination
+	terminated chan struct{} // chan to notify watcher termination status
+}
+
+// NewDeviceWatcher returns a new instance of DeviceWatcher.
+func NewDeviceWatcher() *DeviceWatcher {
+	return &DeviceWatcher{Event: make(chan string), Error: make(chan error),
+		stop: false, terminated: make(chan struct{})}
+}
+
+// Init starts a goroutine that emits device events/errors via watcher channels.
+func (w *DeviceWatcher) Init() {
+	go func() {
+		for !w.stop {
+			log.Info("Waiting for a device change event ")
+			w.awaitEventAndEmit()
+		}
+		close(w.Event)
+		close(w.Error)
+		log.Info("Watcher channels closed ")
+		w.terminated <- struct{}{}
+	}()
+}
+
+// awaitEventAndEmit awaits and emits a device change event or an error.
+func (w *DeviceWatcher) awaitEventAndEmit() {
+	script := fmt.Sprintf(devChangeWaitScript, devChangeTimeout)
+	out, err := exec.Command(powershell, script).Output()
+	if err != nil {
+		log.WithFields(log.Fields{"err": err}).Error("Failed to watch device event ")
+		select {
+		case w.Error <- err:
+			log.WithFields(log.Fields{"err": err}).Info("Successfully emitted error ")
+		default:
+			log.WithFields(log.Fields{"err": err}).Warn("Couldn't emit error, continuing.. ")
+		}
+		return
+	}
+
+	log.WithFields(log.Fields{"out": string(out)}).Info("Watcher script executed ")
+	event := tailSegment(out, lf, 2)
+	select {
+	case w.Event <- event:
+		log.WithFields(log.Fields{"event": event}).Info("Successfully emitted event ")
+	default:
```

Comment:
Since we're sending the event to a `chan`, the flow will immediately drop to the `default` case on noticing a full buffer.